packages/mssql/src/index.ts

```diff
@@ -0,0 +1,403 @@
+        const run = (
+          sql: string,
+          _values?: ReadonlyArray<any>,
+          transform = true,
+          rowsAsArray = false,
+        ) =>
+          Effect.async<never, SqlError, any>(resume => {
+            const req = new Tedious.Request(sql, (error, _rowCount, result) => {
+              if (error) {
+                resume(
+                  Debug.untraced(() => Effect.fail(SqlError(error.message, error))
+                  )
+                )
+              } else if (transform &&
+                !rowsAsArray &&
+                options.transformResultNames) {
+                resume(
+                  Debug.untraced(() => Effect.succeed(transformRows(result)))
+                )
+              } else {
+                resume(Debug.untraced(() => Effect.succeed(result)))
+              }
+            })
+            conn.execSql(req)
+          })
+
+        return {
+          execute(statement) {
+            const [sql, params] = compiler.compile(statement)
+            return run(sql, params)
+          },
+          executeWithoutTransform(statement) {
+            const [sql, params] = compiler.compile(statement)
+            return run(sql, params, false)
+          },
+          executeValues(statement) {
+            const [sql, params] = compiler.compile(statement)
+            return run(sql, params, true, true)
+          },
+          executeRaw(sql, params) {
+            return run(sql, params)
+          },
+          compile(statement) {
+            return Effect.sync(() => compiler.compile(statement))
+          },
+        }
+      }),
+    )
+
+    const pool = yield* _(
+      Pool.makeWithTTL(
+        makeConnection,
+        options.minConnections ?? 1,
+        options.maxConnections ?? 10,
+        options.connectionTTL ?? minutes(45),
+      ),
+    )
+
+    return Object.assign(Client.make(pool.get(), pool.get()), {
+      config: options,
+    })
+  })
+
+/**
+ * @category constructor
+ * @since 1.0.0
+ */
+export const makeLayer = (config: Config.Config.Wrap<TediusClientConfig>) =>
+  Layer.scoped(tag, Effect.flatMap(Effect.config(Config.unwrap(config)), make))
+// UPDATE e
+// SET hire_date = t.hire_date
+// FROM dbo.employee e
+// JOIN (
+//     VALUES
+//         ('PMA42628M', '1979-03-15'),
+//         ('PSA89086M', '1988-12-22')
+// ) t (emp_id, hire_date) ON t.emp_id = e.emp_id
+export const makeCompiler = (transform?: (_: string) => string) => {
+  return Statement.makeCompiler<TediousCustom>(
+    (_,v) => `@${v}`,
+    transform ? _ => escape(transform(_)) : escape,
+    (placeholders, valueAlias, valueColumns, values) => [
+      `(values ${placeholders}) AS ${valueAlias}${valueColumns}`,
+      values.flat(),
+    ],
+    (type, placeholder) => {
+      switch (type.kind) {
+        case "TediusType": {
+          return [placeholder(), [type.i0 as any, type.i1] as any]
+        }
+      }
+    },
+  )
+}
+
+// custom types
+
+type TediousCustom =
```

Comment:
Not sure where the responsibility should lie, left the types and everything here for your consideration

Reply:
Replaced with a lower level `MssqlParam` type.